doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,547 @@
+
+We can change the type of these fields to be `int` instead of `string` and
+define auxiliary `BuiltinTable` sources that are populated with the currently
+available sources on startup.
+
+The current design is simpler and optimizes for speed of delivering the MVP.
+This alternative is a performance optimization that can be done as a follow-up
+if needed.
+
+## Open questions
+
+<!--
+What is left unaddressed by this design document that needs to be
+closed out?
+
+When a design document is authored and shared, there might still be
+open questions that need to be explored. Through the design document
+process, you are responsible for getting answers to these open
+questions. All open questions should be answered by the time a design
+document is merged.
+-->
+
+### RBAC issues
+
+Currently, `Builtin` objects are created with a `sensitivity` field that
+is assigned a `DataSensitivity` value.
+
+```rust
+/// The extent to which data in a builtin object
+/// should be considered "sensitive" and therefore
+/// access to it restricted.
+#[derive(Clone, Debug, Hash, Serialize)]
+pub enum DataSensitivity {
+    /// Any user may query the object.
+    Public,
+    /// Superusers or Materialize staff may query the object.
+    SuperuserAndSupport,
+    /// Only superusers may query the object.
+    Superuser,
+}
+```
+
+Some of the builtin objects proposed in the [Catalog Schema](#catalog-schema)
+(such as `mz_all_optimizer_notices`) are only meant to store raw data and
+support the construction of user-facing builtin views. Those can be all defined
+with sensitivity `DataSensitivity::Superuser`. There are two possible policies
+regarding the sensitivity of the user-facing builtin views of which we need to
+pick one. My proposal is go with either (2) and (3), but ultimately the decision
+should be made by the product and security teams:
+
+1. Use `DataSensitivity::Public`.
+   - ☑ All notices will be visible to all Materialize users.
+   - ☒ The `mz_optimizer_notices` view might leak information about DDL
+     statements in the text fields of notices scoped to these statements (for
+     example the `hint` text might leak a literal constraint from a `WHERE`
+     clause for a `CREATE MATERIALIZED VIEW` statement).
+2. Use `DataSensitivity::SuperuserAndSupport` (permissive).
+   - ☑ All notices will be visible to all Materialize users.
+   - ☒ Notices will be visible to customers only from the UX. The UX will need
+     to be built on top of a connection that uses the `mz_support` user.
+   - ☒ The UX might leak information by listing notices for catalog items that
+     are not owned by the current user with the same implications as above.
+2. Use `DataSensitivity::SuperuserAndSupport` (restrictive).
+   - ☒ All notices will be visible only to Materialize superusers and
+     Materialize support (through the impersonated console).
+   - ☒ Notices will be visible to customers only from the UX. The UX will hide
```

Comment:
Not following why this must be the case? UX seems like the primary interface, but I'm not sure why SQL wouldn't be supported.

Reply:
I didn't consider that regular users where `User::is_external_admin()` is true will also be able to see the notices through the SQL console. I updated the text a bit to reflect that.